packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +400,340 @@ contract AmmPool is IBlockReceiver {
+            require(update.tokenID == ctx.tokens[i].tokenID, "INVALID_TX_DATA");
             require(update.feeBips == feeBips, "INVALID_TX_DATA");
-            require(update.tokenWeight == tokens[i].weight, "INVALID_TX_DATA");
+            require(update.tokenWeight == (start ? 0 : ctx.tokens[i].weight), "INVALID_TX_DATA");
             // Now approve this AMM update
             update.validUntil = 0xffffffff;
-            bytes32 txHash = AmmUpdateTransaction.hashTx(exchangeDomainSeparator, update);
+            bytes32 txHash = AmmUpdateTransaction.hashTx(ctx.exchangeDomainSeparator, update);
             exchange.approveTransaction(address(this), txHash);
-            numTransactionsConsumed++;
-            // AMM account balance now available onchain
-            ammBalancesInAccount[i] = update.balance;
-            ammBalances[i] = ammBalancesInAccount[i];
+            ctx.numTransactionsConsumed++;
+            if (start) {
+                // AMM account balance now available onchain
+                ctx.ammBalancesInAccount[i] = update.balance;
+                ctx.ammBalances[i] = ctx.ammBalancesInAccount[i];
+            }
         }
+    }
 
-        // Process work in the queue
-        for (uint n = queuePos; n < queuePos + auxData.numItems; n++) {
-            QueueItem memory item = queue[n];
-            // TODO: all the necessary pool logic, mint/burn LP tokens,...
-            if (item.workType == WorkType.JOIN) {
-                PoolJoin memory join = item.join;
-                for (uint i = 0; i < tokens.length; i++) {
-                    uint amount = join.maxAmountsIn[i];
-                    ammBalances[i] = ammBalances[i].add(amount);
-                    // Unlock the amount locked for this join
-                    locked[item.owner][tokens[i].addr] = locked[item.owner][tokens[i].addr].sub(join.maxAmountsIn[i]);
-                    // Make the amount unavailable for withdrawing
-                    balance[item.owner][tokens[i].addr] = balance[item.owner][tokens[i].addr].sub(amount);
-                }
-            } else if (item.workType == WorkType.EXIT) {
-                PoolExit memory exit = item.exit;
-                for (uint i = 0; i < tokens.length; i++) {
-                    uint amount = exit.minAmountsOut[i];
-                    ammBalances[i] = ammBalances[i].sub(amount);
-                    // Make the amount available for withdrawing
-                    balance[item.owner][tokens[i].addr] = balance[item.owner][tokens[i].addr].add(amount);
+    function processJoin(
+        Context  memory ctx,
+        PoolJoin memory join,
+        bytes    memory signature
+        )
+        internal
+    {
+        bytes32 poolTxHash = hashPoolJoin(ctx.DOMAIN_SEPARATOR, join);
+        if (signature.length == 0) {
+            require(queue[queuePos].txHash == poolTxHash, "NOT_APPROVED");
+            delete queue[queuePos];
+            queuePos++;
```

Comment:
if we delete the element from the queue, should we still do `queuePos++`?

Reply:
Currently the queue needs to be handled FIFO, so it forces the operator to handle the pool transactions in the correct order. Though I'm not that big of a fan of the queue and would like to remove it.